### ISSUE TITLE
Fuse mask implementation

### ISSUE THREAD

**Issue by karunreddy30** [bzhangGo/zero#3]
Hi,

The mask implementation for aan seems to be different to what you describe in the paper 
https://arxiv.org/pdf/1805.00631.pdf?
Specifically in Figure 3.

Based on Figure 3, I would think the mask computation for sequence length L would be as below: 

mask = tf.matrix_band_part(tf.ones([L, L]), -1, 0)
factor = tf.reciprocal(tf.cumsum(tf.ones_like(mask)))
masked_mask = factor * mask
masked_mask = tf.expand_dims(masked_mask, 0)

The mask computation you did at the following location seems to completely different, in that there is a gate and softmax computation.
https://github.com/bzhangGo/zero/blob/7bb3250e4b804c1ab4f30dc7e9ba7dc51990e4f6/func.py#L390

**Comment by bzhangGo** [comment]
Hi,

Your implementation for AAN mask is more concise. But both functions similarly. Note that: `softmax([1, 1, -1e8, -1e8]) => [0.5, 0.5, 0, 0]`.

**Comment by karunreddy30** [comment]
Apologies if am missing something, but doesn't it have to be `-inf` on Line 396?

With the current setup, `weight = tf.nn.softmax(mask + (1.0 - mask) * inf)`, it would result in `softmax([1, 1, 1e8, 1e8])`

**Comment by bzhangGo** [comment]
Thanks for pointing this out. 

There is a mismatch. I actually set `inf` to -1e8 https://github.com/bzhangGo/zero/blob/7bb3250e4b804c1ab4f30dc7e9ba7dc51990e4f6/func.py#L376 
Will correct it soon.

**Comment by karunreddy30** [comment]
Aah! I should have noticed that. Thanks for getting back.
